Add Ceiling and Floor functions to redblacktree

otnt opened the pull request:
Request:
The redblacktree implementation now is :sparkles: great :sparkles: , and it would be nicer to have Ceiling and Floor functions.

Reasons:
Ceiling and Floor functions are frequently used, and are almost standard functions for tree data structure.

My implementation:
I implemented these two functions in an iterative way. It's time complexity is O(log(N)), and space complexity is O(1).
I also implemented test cases for these two functions.

emirpasic commented:
@otnt 

Please check the following example on how to extend [red-black tree](https://github.com/emirpasic/gods/blob/master/examples/redblacktreeextended.go).

This is the IMHO better approach rather than injecting each requested feature directly into the implementation. See discussion #2

otnt commented:
@emirpasic 

That makes sense. So you think it would be nice to put these two functions in [redblacktreeextended.go](https://github.com/emirpasic/gods/blob/master/examples/redblacktreeextended.go)?

If so, I'll do that in these two days.

emirpasic commented on trees/redblacktree/redblacktree.go:
```diff
@@ -126,6 +126,64 @@ func (tree *Tree) Get(key interface{}) (value interface{}, found bool) {
+// A ceiling node may not be found, either because the tree is empty, or because
+// all nodes in the tree is smaller than the given node.
+// Key should adhere to the comparator's type assertion, otherwise method panics.
+func (tree *Tree) Ceiling(key interface{}) (ceilingKey interface{}, value interface{}, found bool) {
```
This can be simplified to two return parameters: node and found. Node is exposed now and already contains the key and the value.

emirpasic commented on trees/redblacktree/redblacktree.go:
```diff
@@ -126,6 +126,64 @@ func (tree *Tree) Get(key interface{}) (value interface{}, found bool) {
+// A floor node may not be found, either because the tree is empty, or because
+// all nodes in the tree is larger than the given node.
+// Key should adhere to the comparator's type assertion, otherwise method panics.
+func (tree *Tree) Floor(key interface{}) (floorKey interface{}, value interface{}, found bool) {
```
Same as the above comment.

emirpasic commented on trees/redblacktree/redblacktree.go:
```diff
@@ -126,6 +126,64 @@ func (tree *Tree) Get(key interface{}) (value interface{}, found bool) {
+	ceiling := &Node{}
+	found = false
+
+	node := tree.Root
```
We should use pointers here when traversing the tree. (I'll check other parts of code, if I made the same mistake)

emirpasic commented on trees/redblacktree/redblacktree.go:
```diff
@@ -126,6 +126,64 @@ func (tree *Tree) Get(key interface{}) (value interface{}, found bool) {
+	floor := &Node{}
+	found = false
+
+	node := tree.Root
```
Should use pointer (same as above comment):

```go
var node *Tree
node = &tree.Root
```

emirpasic commented:
Sure, I can add that to the examples or you could change this PR? 

This made be realize a blunder of my own, I should have used pointers as an optimization to the tree, so I will try to fix that ASAP.

otnt commented:
@emirpasic 

Thanks for your quick reply.

I moved `Ceiling` and `Floor` functions to `redblacktreeextended.go`. Also, to make comparing nodes possible, I added a `Comparator` function in `redblacktree.go` to get comparator of the red black tree.

Besides, I checked the `Root` of `Tree` is already an `*Node`, so it seems no need to worry that the node is not a pointer.

emirpasic commented on trees/redblacktree/redblacktree.go:
```diff
@@ -192,6 +192,11 @@ func (tree *Tree) Clear() {
 }
 
+// Return comparator of the tree
+func (tree *Tree) Comparator() utils.Comparator {
```
Perhaps there is no need for a getter (conisdered evil nowadays). I've already exposed the _Root_, so simply expose _Compoarator and I can immediately merge the PR:
```go
type Tree struct {
	Root       *Node
	size       int
	Comparator utils.Comparator
}
```

Regarding the pointers, I'll go through them later and see if it could be optimized.

emirpasic commented:
BTW, thanks for contributing!

otnt commented:
No problem! I'm new to open source world, glad to contribute :smile:

emirpasic commented:
Merged and changed some minor cosmetics (exposed Comporator, so no getter)

Cheers
